crates/networking/p2p/rlpx/frame.rs

```diff
@@ -1,36 +1,248 @@
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        let mac_aes_cipher = Aes256Enc::new_from_slice(&self.mac_key.0)?;
+
+        // Receive the message's frame header
+        if src.len() < 32 {
+            // Not enough data to read the frame header.
+            return Ok(None);
+        }
+        let mut frame_header = [0; 32];
+        frame_header.copy_from_slice(&src[..32]);
+
+        // Both are padded to the block's size (16 bytes)
+        let (header_ciphertext, header_mac) = frame_header.split_at_mut(16);
+
+        // Validate MAC header
+        // header-mac-seed = aes(mac-secret, keccak256.digest(egress-mac)[:16]) ^ header-ciphertext
+        let header_mac_seed = {
+            let mac_digest: [u8; 16] = self.ingress_mac.clone().finalize()[..16]
+                .try_into()
+                .map_err(|_| RLPxError::CryptographyError("Invalid mac digest".to_owned()))?;
+            let mut seed = mac_digest.into();
+            mac_aes_cipher.encrypt_block(&mut seed);
+            (H128(seed.into())
+                ^ H128(header_ciphertext.try_into().map_err(|_| {
+                    RLPxError::CryptographyError("Invalid header ciphertext length".to_owned())
+                })?))
+            .0
+        };
+
+        // ingress-mac = keccak256.update(ingress-mac, header-mac-seed)
+        // Use temporary value as it can be discarded if the buffer does not contain yet the full message
+        let mut temp_ingress_mac = self.ingress_mac.clone();
+        temp_ingress_mac.update(header_mac_seed);
+
+        // header-mac = keccak256.digest(egress-mac)[:16]
+        let expected_header_mac = H128(
+            temp_ingress_mac.clone().finalize()[..16]
+                .try_into()
+                .map_err(|_| RLPxError::CryptographyError("Invalid header mac".to_owned()))?,
+        );
+
+        assert_eq!(header_mac, expected_header_mac.0);
+
+        let header_text = header_ciphertext;
+        // Use temporary value as it can be discarded if the buffer does not contain yet the full message
+        let mut temp_ingress_aes = self.ingress_aes.clone();
+        temp_ingress_aes.apply_keystream(header_text);
+
+        // header-data = [capability-id, context-id]
+        // Both are unused, and always zero
+        assert_eq!(&header_text[3..6], &(0_u8, 0_u8).encode_to_vec());
+
+        let frame_size: usize =
+            u32::from_be_bytes([0, header_text[0], header_text[1], header_text[2]])
+                .try_into()
+                .map_err(|_| RLPxError::CryptographyError("Invalid frame size".to_owned()))?;
+
+        let padded_size = frame_size.next_multiple_of(16);
+
+        // Check that the size is not too large to avoid a denial of
+        // service attack where the server runs out of memory.
+        if padded_size > MAX_MESSAGE_SIZE {
+            return Err(RLPxError::InvalidMessageLength());
+        }
+
+        if src.len() < 32 + padded_size + 16 {
```

Comment:
nit: maybe `let total_message_size = 32 + padded_size + 16`? We use it some times and it looks arbitrary at first sight.